### PR TITLE
An instance receiver after a static access is not necessarily non-null

### LINKED ISSUE #553: Static field/method accesses incorrectly set instance receivers to be `@NonNull`
Take this example:

````java
public class Small {
  public static void main(String[] args) {
    Small x = null;
    Object o = x.sfield;
    o = x.field;
  }

  static Object sfield = "";
  Object field = "";
}
````

For a static field/method access, Java only takes the type of the receiver, not the actual value.
So this example does not produce an NPE on the access `x.sfield`, but on the access `x.field`; similarly for method accesses.
The above program passes the Nullness Checker without a warning, meaning this reveals an unsoundness.

The reason is that [method access](https://github.com/eisop/checker-framework/blob/3e84376c96b88949757924897bb61269bc0e0721/checker/src/main/java/org/checkerframework/checker/nullness/NullnessTransfer.java#L347) and [field access](https://github.com/eisop/checker-framework/blob/3e84376c96b88949757924897bb61269bc0e0721/checker/src/main/java/org/checkerframework/checker/nullness/NullnessTransfer.java#L355) transfer functions unconditionally set the receiver to be non-null.
This should only be done if the accessed field/method is an instance field/method, not for static methods/fields.

### PR DESCRIPTION
**Pull request by Ao-senXiong**
Fixes #553

### CODE REVIEW

File: checker/src/main/java/org/checkerframework/checker/nullness/NullnessNoInitTransfer.java [hunk 1]
```diff
@@ -361,9 +361,12 @@ public TransferResult<NullnessNoInitValue, NullnessNoInitStore> visitMethodAcces
             MethodAccessNode n, TransferInput<NullnessNoInitValue, NullnessNoInitStore> p) {
         TransferResult<NullnessNoInitValue, NullnessNoInitStore> result =
                 super.visitMethodAccess(n, p);
-        // In contrast to the conditional makeNonNull in visitMethodInvocation, this
-        // makeNonNull is unconditional, as the receiver is definitely non-null after the access.
-        makeNonNull(result, n.getReceiver());
+        // In contrast to the makeNonNull in visitMethodInvocation, which checks the method is
```
Comment (wmdietl) [inline checker/src/main/java/org/checkerframework/checker/nullness/NullnessNoInitTransfer.java:364]: This comment is rather complicated and inconsistent - before, this was unconditional and the other one was conditional; that distinction no longer exists, so the "in contrast" makes no sense.

How about: `The receiver of an instance method access is non-null. A static method access does not ensure that the receiver is non-null.`

File: checker/src/main/java/org/checkerframework/checker/nullness/NullnessNoInitTransfer.java [hunk 2]
```diff
@@ -372,7 +375,9 @@ public TransferResult<NullnessNoInitValue, NullnessNoInitStore> visitFieldAccess
             FieldAccessNode n, TransferInput<NullnessNoInitValue, NullnessNoInitStore> p) {
         TransferResult<NullnessNoInitValue, NullnessNoInitStore> result =
                 super.visitFieldAccess(n, p);
-        makeNonNull(result, n.getReceiver());
+        if (!n.isStatic()) {
```
Comment (wmdietl) [inline checker/src/main/java/org/checkerframework/checker/nullness/NullnessNoInitTransfer.java:378]: Let's add a comment similar to the one from method access to make this clear.

File: dataflow/src/main/java/org/checkerframework/dataflow/cfg/node/FieldAccessNode.java [hunk 3]
```diff
@@ -24,13 +26,17 @@
  * </pre>
  */
 public class FieldAccessNode extends Node {
-
+    /** The corresponding tree for fieldaccessnode. */
```
Comment (wmdietl) [inline dataflow/src/main/java/org/checkerframework/dataflow/cfg/node/FieldAccessNode.java:29]:
```suggestion
    /** The tree of the field access. */
```

File: dataflow/src/main/java/org/checkerframework/dataflow/cfg/node/FieldAccessNode.java [hunk 4]
```diff
@@ -24,13 +26,17 @@
...
     protected final Tree tree;
+
+    /** The VariableElement. */
```
Comment (wmdietl) [inline dataflow/src/main/java/org/checkerframework/dataflow/cfg/node/FieldAccessNode.java:32]:
```suggestion
    /** The element of the accessed field. */
```

File: dataflow/src/main/java/org/checkerframework/dataflow/cfg/node/FieldAccessNode.java [hunk 5]
```diff
@@ -24,13 +26,17 @@
...
     protected final VariableElement element;
+
+    /** The name of the field accessed by the tree. */
```
Comment (wmdietl) [inline dataflow/src/main/java/org/checkerframework/dataflow/cfg/node/FieldAccessNode.java:35]:
```suggestion
    /** The name of the accessed field. */
```

File: dataflow/src/main/java/org/checkerframework/dataflow/cfg/node/FieldAccessNode.java [hunk 6]
```diff
@@ -24,13 +26,17 @@
...
     protected final String field;
-    protected final Node receiver;
 
-    // TODO: add method to get modifiers (static, access level, ..)
+    /** The receiver node for field being accessed. */
```
Comment (wmdietl) [inline dataflow/src/main/java/org/checkerframework/dataflow/cfg/node/FieldAccessNode.java:38]:
```suggestion
    /** The receiver node of the field access. */
```

File: dataflow/src/main/java/org/checkerframework/dataflow/cfg/node/FieldAccessNode.java [hunk 7]
```diff
@@ -93,7 +99,11 @@ public String toString() {
         return getReceiver() + "." + field;
     }
 
-    /** Is this a static field? */
+    /**
+     * Check the field is a static or not.
+     *
+     * @return a boolean indicates whether the field is static
```
Comment (wmdietl) [inline dataflow/src/main/java/org/checkerframework/dataflow/cfg/node/FieldAccessNode.java:105]:
```suggestion
     * Determine whether the field is static or not.
     *
     * @return whether the field is static or not
```

File: dataflow/src/main/java/org/checkerframework/dataflow/cfg/node/FieldAccessNode.java [hunk 8]
```diff
@@ -118,4 +128,13 @@ public int hashCode() {
     public Collection<Node> getOperands() {
         return Collections.singletonList(receiver);
     }
+
+    /**
+     * Retrieves the modifiers of the method.
+     *
+     * @return the set of modifiers of the method
+     */
+    public Set<Modifier> getMethodModifiers() {
```
Comment (wmdietl) [inline dataflow/src/main/java/org/checkerframework/dataflow/cfg/node/FieldAccessNode.java:137]: Is this method used somewhere? If not, I would remove it.

File: dataflow/src/main/java/org/checkerframework/dataflow/cfg/node/MethodAccessNode.java [hunk 9]
```diff
@@ -22,12 +25,14 @@
  * </pre>
  */
 public class MethodAccessNode extends Node {
-
+    /** The corresponding Expression tree. */
```
Comment (wmdietl) [inline dataflow/src/main/java/org/checkerframework/dataflow/cfg/node/MethodAccessNode.java:28]:
```suggestion
    /** The tree of the method access. */
```

File: dataflow/src/main/java/org/checkerframework/dataflow/cfg/node/MethodAccessNode.java [hunk 10]
```diff
@@ -22,12 +25,14 @@
...
     protected final ExpressionTree tree;
+
+    /** The ExecutatbleElement method. */
```
Comment (wmdietl) [inline dataflow/src/main/java/org/checkerframework/dataflow/cfg/node/MethodAccessNode.java:31]:
```suggestion
    /** The element of the accessed method. */
```

File: dataflow/src/main/java/org/checkerframework/dataflow/cfg/node/MethodAccessNode.java [hunk 11]
```diff
@@ -22,12 +25,14 @@
...
     protected final ExecutableElement method;
-    protected final Node receiver;
 
-    // TODO: add method to get modifiers (static, access level, ..)
+    /** The receiver node for method being accessed. */
```
Comment (wmdietl) [inline dataflow/src/main/java/org/checkerframework/dataflow/cfg/node/MethodAccessNode.java:34]:
```suggestion
    /** The receiver node of the method access. */
```

File: dataflow/src/main/java/org/checkerframework/dataflow/cfg/node/MethodAccessNode.java [hunk 12]
```diff
@@ -90,4 +95,22 @@ public int hashCode() {
     public Collection<Node> getOperands() {
         return Collections.singletonList(receiver);
     }
+
+    /**
+     * Retrieves the modifiers of the method.
+     *
+     * @return the set of modifiers of the method
+     */
+    public Set<Modifier> getMethodModifiers() {
```
Comment (wmdietl) [inline dataflow/src/main/java/org/checkerframework/dataflow/cfg/node/MethodAccessNode.java:104]: Remove unused method.

File: dataflow/src/main/java/org/checkerframework/dataflow/cfg/node/MethodAccessNode.java [hunk 13]
```diff
@@ -90,4 +95,22 @@ public int hashCode() {
...
+        return method.getModifiers();
+    }
+
+    /**
+     * Check the method is a static or not.
```
Comment (wmdietl) [inline dataflow/src/main/java/org/checkerframework/dataflow/cfg/node/MethodAccessNode.java:109]: Same changes as for field access.
We usually use "check" if the method does some checks and raises errors. Here you get a boolean result.

File: docs/CHANGELOG.md [hunk 14]
```diff
@@ -32,8 +32,7 @@ Changed the return types of
 
 **Closed issues:**
 
-eisop#297, eisop#376, eisop#400, eisop#532, typetools#1590.
-
+eisop#297, eisop#376, eisop#400, eisop#532 eisop#533, typetools#1590.
```
Comment (wmdietl) [inline docs/CHANGELOG.md:35]:
```suggestion
eisop#297, eisop#376, eisop#400, eisop#532, eisop#533, typetools#1590.
```

**Review by wmdietl — CHANGES_REQUESTED** [review]
A bunch of rewordings that you can just apply, plus a few questions.
Thanks!

File: checker/src/main/java/org/checkerframework/checker/nullness/NullnessNoInitTransfer.java (hunk 1 above)
Comment (Ao-senXiong) [reply checker/src/main/java/org/checkerframework/checker/nullness/NullnessNoInitTransfer.java:364]: Ok. I will add change to this sentence.

File: dataflow/src/main/java/org/checkerframework/dataflow/cfg/node/FieldAccessNode.java (hunk 8 above)
Comment (Ao-senXiong) [reply dataflow/src/main/java/org/checkerframework/dataflow/cfg/node/FieldAccessNode.java:137]: ok, I will remove it.

File: dataflow/src/main/java/org/checkerframework/dataflow/cfg/node/MethodAccessNode.java (hunk 12 above)
Comment (Ao-senXiong) [reply dataflow/src/main/java/org/checkerframework/dataflow/cfg/node/MethodAccessNode.java:104]: Ok.

File: dataflow/src/main/java/org/checkerframework/dataflow/cfg/node/MethodAccessNode.java (hunk 13 above)
Comment (Ao-senXiong) [reply dataflow/src/main/java/org/checkerframework/dataflow/cfg/node/MethodAccessNode.java:98]: Done.

File: docs/CHANGELOG.md (hunk 14 above)
Comment (Ao-senXiong) [reply docs/CHANGELOG.md]: I add `fix a bug` sentence just like old commits

**Pull request merged**